File: Services/Interfaces/web_interface/controllers/community.py
#  Drakkar-Software OctoBot-Interfaces
#  Copyright (c) Drakkar-Software, All rights reserved.
#
#  This library is free software; you can redistribute it and/or
#  modify it under the terms of the GNU Lesser General Public
#  License as published by the Free Software Foundation; either
#  version 3.0 of the License, or (at your option) any later version.
#
#  This library is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#  Lesser General Public License for more details.
#
#  You should have received a copy of the GNU Lesser General Public
#  License along with this library.

import flask

import octobot_commons.authentication as authentication
import tentacles.Services.Interfaces.web_interface as web_interface
import tentacles.Services.Interfaces.web_interface.login as login
import tentacles.Services.Interfaces.web_interface.models as models


@web_interface.server_instance.route("/community")
@login.login_required_when_activated
def community():
    authenticator = authentication.Authenticator.instance()
    logged_in_email = None
    use_preview = not authenticator.can_authenticate()
    try:
        models.wait_for_login_if_processing()
        logged_in_email = authenticator.get_logged_in_email()
    except (authentication.AuthenticationRequired, authentication.UnavailableError):
        pass
    except Exception as e:
        flask.flash(f"Error when contacting the community server: {e}", "error")
    if logged_in_email is None and not use_preview:
        return flask.redirect('community_login')
    tentacles_packages = models.get_account_tentacles_packages(authenticator) if logged_in_email else []
    default_image = flask.url_for('static', filename="img/community/tentacles_packages_previews/octobot.png")
    return flask.render_template(
        'community.html',
        use_preview=use_preview,
        preview_tentacles_packages=models.get_preview_tentacles_packages(flask.url_for),
        current_logged_in_email=logged_in_email,
        role=authenticator.user_account.supports.support_role,
        is_donor=bool(authenticator.user_account.supports.is_donor()),
        tentacles_packages=tentacles_packages,
        current_bots_stats=models.get_current_octobots_stats(),
        all_user_bots=models.get_all_user_bots(),
        selected_user_bot=models.get_selected_user_bot(),
        default_tentacles_package_image=default_image,
        can_logout=models.can_logout(),
        can_select_bot=models.can_select_bot(),
    )


@web_interface.server_instance.route("/community_metrics")
@login.login_required_when_activated
def community_metrics():
    can_get_metrics = models.can_get_community_metrics()
    display_metrics = models.get_community_metrics_to_display() if can_get_metrics else None
    return flask.render_template('community_metrics.html',
                                 can_get_metrics=can_get_metrics,
                                 community_metrics=display_metrics
                                 )

File: Services/Interfaces/web_interface/static/js/components/community.js
/*
 * Drakkar-Software OctoBot
 * Copyright (c) Drakkar-Software, All rights reserved.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3.0 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library.
 */

function disablePackagesOperations(should_lock=true){
    const disabled_attr = 'disabled';
    $("#synchronize-tentacles").prop(disabled_attr, should_lock);
    $(".install-package-button").prop(disabled_attr, should_lock);
}

function syncPackages(source){
    const update_url = source.attr(update_url_attr);
    disablePackagesOperations();
    send_and_interpret_bot_update({}, update_url, source, packagesOperationSuccessCallback, packagesOperationErrorCallback);
}

function reloadTable(){
    $('.table').each(function () {
        $(this).DataTable();
    });
    registerPackagesEvents();
}

function registerPackagesEvents(){
    $(".install-package-button").click(function (){
        const element = $(this);
        const update_url = element.attr(update_url_attr);
        const data = {
            "url": element.data("package-url"),
            "version": element.data("package-latest-compatible-version")
        };
        disablePackagesOperations();
        send_and_interpret_bot_update(data, update_url, element, packagesOperationSuccessCallback, packagesOperationErrorCallback);
    });
}

function reloadOwnedPackages(){
    $("#owned-tentacles").load(location.href + " #owned-tentacles", function(){
        reloadTable();
    });
}

function packagesOperationSuccessCallback(updated_data, update_url, dom_root_element, msg, status){
    disablePackagesOperations(false);
    reloadOwnedPackages();
    create_alert("success", "Packages operation succeed", msg);
}

function packagesOperationErrorCallback(updated_data, update_url, dom_root_element, result, status, error){
    disablePackagesOperations(false);
    reloadOwnedPackages();
    create_alert("error", "Error when managing packages: "+result.responseText, "");
}

function displayBotSelectorWhenNoSelectedBot(){
    if($("#bot-selector").find("button[data-role='selected-bot']").length === 0) {
        // no selected bot, force selection
        $('#bot-select-modal').modal({backdrop: 'static', keyboard: false})
    }
}

function disableBotsSelectAndCreate(disabled){
    $("#bot-selector").find("button[data-role='select-bot']").attr("disabled", disabled);
    $("#create-new-bot").attr("disabled", disabled);
}

function initBotsCallbacks(){
    $("#bot-selector").find("button[data-role='select-bot']").click((element) => {
        const selectButton = $(element.target);
        const data = selectButton.data("bot-id")
        disableBotsSelectAndCreate(true);
        selectButton.html("<i class='fa fa-spinner fa-spin'></i>")
        const update_url = $("#bot-selector").data("update-url");
        send_and_interpret_bot_update(data, update_url, null,
            botOperationSuccessCallback, botOperationErrorCallback);

    })
    $("#create-new-bot").click((element) => {
        const createButton = $(element.target);
        const update_url = createButton.data("update-url");
        disableBotsSelectAndCreate(true);
        createButton.html("<i class='fa fa-spinner fa-spin'></i> Creating ...")
        send_and_interpret_bot_update({}, update_url, null,
            botOperationSuccessCallback, botOperationErrorCallback);
    })
}

function botOperationSuccessCallback(updated_data, update_url, dom_root_element, result, status, error){
    // reload the page to retest bots
    window.location.reload();
}

function botOperationErrorCallback(updated_data, update_url, dom_root_element, result, status, error){
    create_alert("error", "Error when managing bots: "+result.responseText, "");
}

function initLoginSubmit(){
    $("form[name=community-login]").on("submit", () => {
        $("input[value=Login]").attr("disabled", true);
    });
}

$(document).ready(function() {
    reloadTable();
    $("#synchronize-tentacles").click(function(){
        syncPackages($(this));
    });
    displayBotSelectorWhenNoSelectedBot();
    initBotsCallbacks();
    initLoginSubmit();
});
